### PR TITLE
explicitly specify envs to be exported in function

### PR DESCRIPTION
**Pull request by Shell32-Natsu**
part of https://github.com/GoogleContainerTools/kpt/issues/951

This PR only contains the changes in kyaml

### CODE REVIEW

File: kyaml/fn/runtime/container/container_test.go [hunk 1]
```diff
@@ -137,20 +136,13 @@ metadata:
 			}
 			tt.instance.Exec.FunctionConfig = cfg
 
-			os.Setenv("KYAML_TEST", "FOO")
-			tt.instance.setupExec()
+			tt.instance.Envs.AddKeyValue("KYAML_TEST", "FOO")
+			tt.expectedArgs = append(tt.expectedArgs, tt.instance.Envs.GetDockerFlags()...)
```
Comment (monopole) [inline kyaml/fn/runtime/container/container_test.go:140]: formatting

File: kyaml/fn/runtime/runtimeutil/functiontypes.go [hunk 2]
```diff
@@ -33,6 +34,78 @@ const (
 	UserNobody ContainerUser = "nobody"
 )
 
+const defaultEnvValue string = "true"
+
+// ContainerEnvs contains the environment variables for the container
```
Comment (monopole) [inline kyaml/fn/runtime/runtimeutil/functiontypes.go:39]: It's _one_ environment, with multiple environment variables, flags, etc. defined inside.
suggest:
```
// ContainerEnv defines the environment present in a container.
```

File: kyaml/fn/runtime/runtimeutil/functiontypes.go [hunk 3]
```diff
@@ -33,6 +34,78 @@ const (
...
+type ContainerEnvs struct {
+	// EnvsMap is a key-value map that will be set as env in container
+	EnvsMap map[string]string `json:"envsMap,omitempty" yaml:"envsMap,omitempty"`
```
Comment (monopole) [inline kyaml/fn/runtime/runtimeutil/functiontypes.go:42]: EnvVars

(don't include the type in the var name, e.g. "NameString", "AgeInt")

File: kyaml/fn/runtime/runtimeutil/functiontypes.go [hunk 4]
```diff
@@ -33,6 +34,78 @@ const (
...
+// GetDockerFlags returns docker run style env flags
+func (ce *ContainerEnvs) GetDockerFlags() []string {
+	envs := ce.EnvsMap
+	if envs == nil {
+		envs = make(map[string]string)
+	}
+
+	// default envs
+	envs["LOG_TO_STDERR"] = defaultEnvValue
+	envs["STRUCTURED_RESULTS"] = defaultEnvValue
```
Comment (monopole) [inline kyaml/fn/runtime/runtimeutil/functiontypes.go:57]: Line 50 is a map ptr copy, not a map copy, so these lines mutate the instance.  

A method named `GetWhatever` should not be a mutator.

To set up defaults, define a factory function like

 func MakeContainerEnv() *ContainerEnv {
   // install defaults.
}

File: kyaml/fn/runtime/runtimeutil/functiontypes.go [hunk 5]
```diff
@@ -33,6 +34,78 @@ const (
...
 )
 
+const defaultEnvValue string = "true"
+
+// ContainerEnvs contains the environment variables for the container
+type ContainerEnvs struct {
+	// EnvsMap is a key-value map that will be set as env in container
+	EnvsMap map[string]string `json:"envsMap,omitempty" yaml:"envsMap,omitempty"`
+
+	// ExportKeys are only env key. Value will be the value in the host system
```
Comment (monopole) [inline kyaml/fn/runtime/runtimeutil/functiontypes.go:44]: Are these a subset of the keys in EnvVars?

Comment (Shell32-Natsu) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:44]: No, EnvsVar contains key-value map. ExportedKeys contains the env key that user wants to expose to container from host system. No value specified.

File: kyaml/fn/runtime/runtimeutil/functiontypes.go (hunk 4 above)
Comment (Shell32-Natsu) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:57]: There is a factory function in line 103. Will add the default values there.

Comment (Shell32-Natsu) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:57]: I am still not very sure about which is better in a factory method, return a pointer or a value?

File: kyaml/fn/runtime/runtimeutil/functiontypes.go (hunk 2 above)
Comment (monopole) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:39]: What i mean here - i don't think the struct's name should have an s on the end.

File: kyaml/fn/runtime/runtimeutil/functiontypes.go (hunk 5 above)
Comment (monopole) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:44]: So maybe call this `VarsToExport`.

File: kyaml/fn/runtime/runtimeutil/functiontypes.go (hunk 4 above)
Comment (monopole) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:57]: I don't think we've been strict here.

I want to say return values for "small", immutable structs (like this one seems to be) and pointers for mutable structs or for structs so big that coping would be a problem.  

But on the whole, kustomize typically passes around pointers.

File: kyaml/fn/runtime/runtimeutil/functiontypes.go [hunk 6]
```diff
@@ -75,6 +161,12 @@ type ContainerSpec struct {
 
 	// User is the username/uid that application runs as in continer
 	User ContainerUser `json:"user,omitempty" yaml:"user,omitempty"`
+
+	// EnvRaw is a slice of env string.
+	EnvRaw []string `json:"envs,omitempty" yaml:"envs,omitempty"`
```
Comment (monopole) [inline kyaml/fn/runtime/runtimeutil/functiontypes.go:166]: Provide a better comment.  Is this to provide a means to serialize the contents of ContainerEnv?

Comment (Shell32-Natsu) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:166]: Yes, it's used to serialize/deserialize envs from function configs.

Comment (monopole) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:166]: We shouldn't have two representations of the same data.

Drop this slice, and add field tags to ContainerEnv instead.

Examples: https://github.com/kubernetes-sigs/kustomize/blob/master/kyaml/yaml/types.go#L166

Comment (Shell32-Natsu) [reply kyaml/fn/runtime/runtimeutil/functiontypes.go:166]: But the envs should be a slice of string in the function config yaml. Can I serialize the ContainerEnv struct into a slice of string by tag?

**Pull request merged**